base/Base/Base_Base_implementation.F90

```diff
@@ -3169,6 +3169,144 @@ end function ijsearch
 
   end subroutine MAPL_GetHorzIJIndex
 
+  module subroutine MAPL_GetGlobalHorzIJIndex(npts,II,JJ,lon,lat,lonR8,latR8,Grid, rc)
+    implicit none
+    !ARGUMENTS:
+    integer,                      intent(in   ) :: npts ! number of points in lat and lon arrays
+    integer,                      intent(inout) :: II(npts) ! array of the first index for each lat and lon
+    integer,                      intent(inout) :: JJ(npts) ! array of the second index for each lat and lon
+    real, optional,               intent(in   ) :: lon(npts) ! array of longitudes in radians
+    real, optional,               intent(in   ) :: lat(npts) ! array of latitudes in radians
+    real(ESMF_KIND_R8), optional, intent(in   ) :: lonR8(npts) ! array of longitudes in radians
+    real(ESMF_KIND_R8), optional, intent(in   ) :: latR8(npts) ! array of latitudes in radians
+    type(ESMF_Grid),    optional, intent(inout) :: Grid ! ESMF grid
+    integer,            optional, intent(out  ) :: rc  ! return code
+
+    integer :: status
+    integer :: dims(3), IM_WORLD, JM_WORLD
+    real(ESMF_KIND_R8), allocatable, dimension (:,:) :: xyz
+    real(ESMF_KIND_R8), allocatable, dimension (:)   :: x,y,z
+    real(ESMF_KIND_R8), allocatable :: max_values(:)
+    real(ESMF_KIND_R8)              :: sqr2, alpha, dalpha, error, shift
+    real(ESMF_KIND_R8), allocatable :: lons(:), lats(:)
+
+!
+! TODO
+! 
+!   make sure the grid is generated with gnomonic_ed, i.e., equal distance of the edge on the sphere surface
+
+    if (npts == 0 ) then
+      _RETURN(_SUCCESS)
+    endif
+
+    if ( .not. present(grid)) then
+      _ASSERT(.false., "need cubed-sphere grid")
```

Comment:
Hmm. If this is true, should we have a more robust check here? Something that would trigger if a user passed in a lat-lon, tripolar, or EASE grid?

Or is `grid` something that only happens with cube-sphere?

Reply:
Yes. Actually, we need more details here. Not all cubed-sphere grid can be done in this way.

Reply:
add the test of the grid at the very begining